src/parachain/issue.ts

```diff
@@ -113,40 +154,128 @@ export class DefaultIssueAPI implements IssueAPI {
+                singleVault = BN.max(singleVault, vaultAvailable) as PolkaBTC;
+                return [singleVault, maxTotal];
+            },
+            [new BN(0) as PolkaBTC, new BN(0) as PolkaBTC]
+        );
+        return { singleVaultMaxIssuable, totalMaxIssuable };
+    }
+
     /**
      * @param events The EventRecord array returned after sending an issue request transaction
      * @returns The issueId associated with the request. If the EventRecord array does not
      * contain issue request events, the function throws an error.
      */
-    private getRequestIdFromEvents(events: EventRecord[]): Hash {
+    private getRequestIdsFromEvents(events: EventRecord[]): Hash[] {
+        const ids = new Array<Hash>();
         for (const { event } of events) {
             if (this.api.events.issue.RequestIssue.is(event)) {
                 const hash = this.api.createType("Hash", event.data[0]);
-                return hash;
+                ids.push(hash);
             }
         }
+        if (ids.length > 0) return ids;
         throw new Error("Request transaction failed");
     }
 
-    async request(amountSat: PolkaBTC, vaultId?: AccountId): Promise<IssueRequestResult> {
+    private allocateAmountsToVaults(
+        vaultsWithIssuableTokens: Map<AccountId, PolkaBTC>,
+        amountToAllocate: PolkaBTC
+    ): Map<AccountId, PolkaBTC> {
+        const maxReservationPercent = 100; // don't reserve more than 90% of a vault's collateral
+        const allocations = new Map<AccountId, PolkaBTC>();
+        // iterable array in ascending order of issuing capacity:
+        const vaultsArray = [...vaultsWithIssuableTokens.entries()].reverse().map((entry) => [entry[0],
+            entry[1].divn(100).muln(maxReservationPercent)] as [AccountId, PolkaBTC]);
+        console.log(`AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAa\n starting to allocate: ${amountToAllocate.toString()}`);
+        while (amountToAllocate.gtn(0)) {
+            // find first vault that can fulfil request (or undefined if none)
+            const firstSuitable = vaultsArray.findIndex(([_, available]) => available.gte(amountToAllocate));
+            console.log(`index ${firstSuitable} will be used`);
+            let vault, amount;
+            if (firstSuitable !== -1) {
+                // at least one vault can fulfil in full
+                // select random vault able to fulfil request
+                const range = vaultsArray.length - firstSuitable;
+                const idx = Math.floor(Math.random() * range) + firstSuitable;
+                console.log(`Selecting vault at index ${idx}`);
+                vault = vaultsArray[idx][0];
+                amount = amountToAllocate;
+                console.log(`Allocating ${amount} to ${vault} all at once`);
+            } else {
+                // else allocate greedily
+                if (vaultsArray.length === 0) throw new Error("Insufficient capacity to fulfil request");
```

Comment:
that should likely be thrown earlier or the check here should be if length < 2 or so

Reply:
Good idea, I'll add another check at the start.
I'll still keep this one in to make the type safety obvious.